### ISSUE TITLE
Improve auto-completion of DOM tags in React. 

### ISSUE THREAD

**Issue by path-of-reason** [zed-industries/zed#4271]
## Description

Currently, when I type "div" in a tsx or jsx file, there is no IntelliSense support. 
Suggestions only work after I type the angle bracket '<', and I still have to close the tag manually, which is inconvenient.

**Comment by emersonjds** [comment]
@path-of-reason This really helps a lot, rewriting tags all the time is tiring

**Comment by emersonjds** [comment]
@JosephTLyons This is a feature that brings a totally different experience to the developer, let's look more at the experience to build a cutting-edge product

**Comment by KoHorizon** [comment]
I agree, this is the only issue that stop me from using this IDE. Might implement it myself, if it's not the priority.

**Comment by chungweileong94** [comment]
Coming from #5267, the tag auto-close doesn’t seem to work for HTML either.

**Comment by Moshyfawn** [comment]
> Coming from #5267, the tag auto-close doesn’t seem to work for HTML either.

Which version of Zed are you on?

**Comment by chungweileong94** [comment]
> > Coming from #5267, the tag auto-close doesn’t seem to work for HTML either.
> 
> Which version of Zed are you on? Do you have the HTML extension installed?

I just tested on the latest preview 0.138.1, and the problem still persists. And I do have the HTML extension installed as well.

https://github.com/zed-industries/zed/assets/15154097/3b56a89e-1925-414e-aebd-237878e0ee8e

**Comment by Moshyfawn** [comment]
It's after you complete the tag with `>` that the LSP closes it for you.

**Comment by chungweileong94** [comment]
> It's after you complete the tag with `>` that the LSP closes it for you.

Ah, my bad, I'm actually showing the wrong demo here. But yeah, but yeah, it's working in the latest 0.138 now😅 
But it wasn't the case for 0.137 before.

**Comment by IlirEdis** [comment]
This keeps me from switching to Zed! Also shortcuts for code snippets such as `rsfc` to create a React function component, etc... This makes your slogan obsolete ;)

**Comment by leognmotta** [comment]
I don't know there is aways something basic that keeps me from switching to zed

**Comment by monarcode** [comment]
This and no snippets support are the main reason I have not switched to Zed fully

typing rfc and getting a react component scaffolded for you is way to good to pass up on

**Comment by osiewicz** [comment]
For what it's worth, we've recently landed support for snippets in https://github.com/zed-industries/zed/pull/13253

**Comment by monarcode** [comment]
@osiewicz Hi, i did see the update on snippet support and already tried it out, how ever, i noticed that the formatting is seemingly bugged.

when the snipper is expanded it does not appear as it should..see screenshots bellow
![Monarch Twitter Media](https://github.com/zed-industries/zed/assets/61241222/e40ecb61-b46e-47d9-a17b-c31cd6cec9a3)
![Monarch Twitter Media (1)](https://github.com/zed-industries/zed/assets/61241222/363a5cde-e6e1-42ef-99c5-77fa8bd2fc1b)

**Comment by osiewicz** [comment]
Do you mean that `<p></p>` is getting de-indented?

**Comment by monarcode** [comment]
Yes, that

**Comment by monarcode** [comment]
Any way to fix that? @osiewicz

**Comment by osiewicz** [comment]
No, I'm sorry, I haven't looked into it yet. A wild guess is that formatting gets applied to completions automatically, which may not be desired for snippets. If so, we should add a way for a language server to opt out of automatic formatting

**Comment by tpolito** [comment]
Any movement on this? I'm on Zed 0.143.7, with the extension and tags aren't closing. 

https://github.com/user-attachments/assets/0b0ce348-448f-4032-b7ac-6fb9beb6caee

**Comment by osiewicz** [comment]
Hey, this is a duplicate of: https://github.com/zed-industries/zed/issues/13762; I'm gonna close out this issue in favour of the other one.

**Comment by feldrok** [comment]
Any news on this?

**Comment by MK732** [comment]
Just switched over to zed, I primarily work in Next.JS so a lot of TSX files. Any update on if this is being fixed soon?

**Comment by akmaldira** [comment]
> Just switched over to zed, I primarily work in Next.JS so a lot of TSX files. Any update on if this is being fixed soon?

same here, just switched to zed today, but i have 2 problem
1. html tag not auto-completion 
2. cmd + / on tsx or jsx not working as well

**Comment by akmaldira** [comment]
@MK732 you must try this

https://github.com/zed-industries/zed/discussions/6666#discussioncomment-10047888

**Comment by TeoBale** [comment]
I think improving the user experience of the default configuration for creating react applications is important to facilitate a conversion to zed...

I've been trying to figure out how to fix a lot of these little problems for a while now, chief among them being the auto-completion of tags within jsx and the order in which the completion tab suggestions are displayed

**Comment by paularmstrong** [comment]
One thing to remember, JSX is really what's important here, not just React. There are quite a few languages that use JSX or some HTML/XML-like markup. All of which tend to have similar needs in terms of autocompletion.

**Comment by dogezen** [comment]
> One thing to remember, JSX is really what's important here, not just React. There are quite a few languages that use JSX or some HTML/XML-like markup. All of which tend to have similar needs in terms of autocompletion.

Agreed, elixir HEEX also requires auto-close feature for HTML tags

**Comment by tealsnow** [comment]
I was very excited to work on a new project with zed, but this issue alone has lead be back to vscode.
Please, someone work on this!

**Comment by mannuch** [comment]
Yeah, auto-closing HTML tags is still not working for me. 

Zed v0.157.5, HTML extension is installed

**Comment by kaungmyathein12** [comment]
any updates on this?

**Comment by librz** [comment]
I also found that even for html files, this doesn't work on vim mode. Using latest version(`v.163.2`) which just shipped today.

**Comment by librz** [comment]
> I also found that even for html files, this doesn't work in vim mode. Using latest version(`v.163.2`) which just shipped today.

One workaround is using multiple select. Like in vs code, holding down `option` then click:

https://github.com/user-attachments/assets/d169718d-158a-4285-b4e5-0b2ecd9a7aa3

Collapse the tag if there's too much in between.

Also vim's shortcut for change insdie word is `ciw`

**Comment by devdumpling** [comment]
+1 still having issues with this in `.astro` files using JSX (or their flavor of JSX). Emmet does help a lot. 

I am using the Astro extension.

**Comment by shizukushq** [comment]
I tried the [solution](https://github.com/zed-industries/zed/issues/5267#issuecomment-2566243172) from @3ux1n3, it helped me, but it doesn't really work with React components. Any news on this?

**Comment by lohenyumnam** [comment]
This is a deal-breaker for me.
